### PR TITLE
Evaluating when conditions

### PR DESCRIPTION
**Pull request by sbarker2**
tests cases are described here:
https://hello.atlassian.net/wiki/spaces/~70121933eac19934744898ba2fcc1c8be3c9c/pages/3203505791/Nadel+Conditional+Hydration+Test+Cases+-+Brainstorm

### CODE REVIEW

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        "data": {
+          "foo": {
+            "hydration__bar__barId": "barId",
+            "__typename__hydration__bar": "Foo",
+            "hydration__bar__type": "someType"
+          }
+        },
+        "extensions": {}
+      }
+  - serviceName: "service2"
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml:97]: Remove this as there is never a hydration.

Right now the test is ambiguous, did the hydration occur and nothing was resolved? or did we not perform a hydration because the condition was false?

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    
+    type Query {
+      barById(id: ID): Bar
+    }
+  service1: |
+    type Foo {
+      barId: ID
+      fooDetails: FooDetails
+      id: ID
+    type: String
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml:54]:
```suggestion
      type: String
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      type: String
+    }
+    
+    type Query {
+      barById(id: ID): Bar
+    }
+  service1: |
+    type Foo {
+      barId: ID
+      fooDetails: FooDetails
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml:52]:
```suggestion
```
Not used

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      id: ID
+    type: String
+    }
+    
+    type FooDetails {
+      externalBarId: ID
+    }
+
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml:60]:
```suggestion
```

Not used

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "someOtherRandomTypeThatNobodyUses" }
+          }
+        }
+      )
+      barLongerInput: Bar @hydrated(service: "service2" field: "barById" arguments: [{name: "id" value: "$source.fooDetails.externalBarId"}])
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml:36]:
```suggestion
```

Not used

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+          }
+        },
+        "extensions": {}
+      }
+# language=JSON
+response: |-
+  {
+    "data": {
+      "foo": {
+        "bar": null
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:122]: Why is this null?

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "someType" }
+          }
+        }
+      )
+      barLongerInput: Bar @hydrated(service: "service2" field: "barById" arguments: [{name: "id" value: "$source.fooDetails.externalBarId"}])
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:36]:
```suggestion
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    
+    type Query {
+      barById(id: ID): Bar
+    }
+  service1: |
+    type Foo {
+      barId: ID
+      fooDetails: FooDetails
+      id: ID
+    type: String
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:54]:
```suggestion
      type: String
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+      type: String
+    }
+    
+    type Query {
+      barById(id: ID): Bar
+    }
+  service1: |
+    type Foo {
+      barId: ID
+      fooDetails: FooDetails
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:52]:
```suggestion
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      id: ID
+    type: String
+    }
+    
+    type FooDetails {
+      externalBarId: ID
+    }
+
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:60]:
```suggestion
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+          }
+        }
+      variables: { }
+    # language=JSON
+    response: |-
+      {
+        "data": {
+          "barById": {
+            "name": "Bar1",
+            "type": "someType"
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:112]:
```suggestion
            "name": "Bar1"
```

The query never asks for  `type`

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        "extensions": {}
+      }
+# language=JSON
+response: |-
+  {
+    "data": {
+      "foo": {
+        "bar": {
+          "name": "Bar1",
+          "type": "someType"
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-condition.yml:124]:
```suggestion
          "name": "Bar1"
```

Ditto

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml (hunk 1 above)
Comment (sbarker2) [reply test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-condition.yml]: good idea. done

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 13]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
...
-        val fieldPath: NadelQueryPath,
-        val value: Any,
+    abstract fun evaluate(fieldValue: Any?): Boolean
+
+    data class StringResultEquals(
+        override val fieldPath: NadelQueryPath,
+        val value: String,
+    ) : NadelHydrationWhenCondition() {
+        override fun evaluate(fieldValue: Any?): Boolean {
+            if (fieldValue is String){
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:15]:
```suggestion
            if (fieldValue is String) {
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 14]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
...
+                return  fieldValue == value
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:16]:
```suggestion
                return fieldValue == value
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 15]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
...
+            }
+            return false
+        }
+    }
+    data class LongResultEquals(
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:21]:
```suggestion

    data class LongResultEquals(
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 16]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
...
+        override val fieldPath: NadelQueryPath,
+        val value: Long,
     ) : NadelHydrationWhenCondition() {
-        override fun evaluate(resultId: String): Boolean {
-            TODO("Not yet implemented")
+        override fun evaluate(fieldValue: Any?): Boolean {
+            if (fieldValue is Int) {
+                return  fieldValue.toLong() == value
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:27]:
```suggestion
                return fieldValue.toLong() == value
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 17]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
...
+            }
+            if (fieldValue is Long) {
+                return  fieldValue == value
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:30]:
```suggestion
                return fieldValue == value
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 18]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
...
     data class StringResultMatches(
-        val fieldPath: NadelQueryPath,
+        override val fieldPath: NadelQueryPath,
         val regex: Regex,
     ) : NadelHydrationWhenCondition() {
-        override fun evaluate(resultId: String): Boolean {
-            TODO("Not yet implemented")
+        override fun evaluate(fieldValue: Any?): Boolean {
+            if (fieldValue is String) {
+                return fieldValue.matches(regex)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:42]: I do wonder if we should use `.contains(Regex)` instead or not.

(Nothing to action for now).

File: lib/src/main/java/graphql/nadel/engine/transform/hydration/NadelHydrationTransform.kt [hunk 19]
```diff
@@ -279,11 +279,21 @@ internal class NadelHydrationTransform(
         hooks: NadelExecutionHooks,
         parentNode: JsonNode,
     ): NadelHydrationFieldInstruction? {
-        return hooks.getHydrationInstruction(
-            instructions,
-            parentNode,
-            state.aliasHelper,
-            state.executionContext.userContext
-        )
+        if (instructions.any{it.condition == null}){
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/transform/hydration/NadelHydrationTransform.kt:282]:
```suggestion
        if (instructions.any { it.condition == null }) {
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt [hunk 20]
```diff
@@ -3,32 +3,57 @@ package graphql.nadel.engine.blueprint.hydration
 import graphql.nadel.engine.transform.query.NadelQueryPath
 
 sealed class NadelHydrationWhenCondition {
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/hydration/NadelHydrationWhenCondition.kt:5]:
```suggestion
internal sealed class NadelHydrationWhenCondition {
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-equals-condition.yml [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "$source.barId"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "someOtherRandomTypeThatNobodyUses" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-equals-condition.yml:32]:
```suggestion
            sourceField: "type"
            predicate: { equals: "someOtherRandomTypeThatNobodyUses" }
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-matches-condition.yml [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "$source.barId"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { matches: "[A-Z]+Type" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-matches-condition.yml:32]:
```suggestion
            sourceField: "type"
            predicate: { matches: "[A-Z]+Type" }
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-starts-with-condition.yml [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "$source.barId"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { startsWith: "hehe" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-false-starts-with-condition.yml:32]:
```suggestion
            sourceField: "type"
            predicate: { startsWith: "hehe" }
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-equals-condition.yml [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "$source.barId"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "someType" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-equals-condition.yml:32]:
```suggestion
            sourceField: "type"
            predicate: { equals: "someType" }
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-matches-condition.yml [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "$source.barId"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { matches: "[a-z]+Type" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-matches-condition.yml:32]:
```suggestion
            sourceField: "type"
            predicate: { matches: "[a-z]+Type" }
```

File: test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-starts-with-condition.yml [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "$source.barId"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { startsWith: "some" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/basic-conditional-hydration-with-true-starts-with-condition.yml:32]:
```suggestion
            sourceField: "type"
            predicate: { startsWith: "some" }
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-equals-predicate-works-with-integer-type.yml [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_A"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: 1 }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-equals-predicate-works-with-integer-type.yml:34]:
```suggestion
            sourceField: "type"
            predicate: { equals: 1 }
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-equals-predicate-works-with-integer-type.yml [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_B"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: 2 }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-equals-predicate-works-with-integer-type.yml:50]:
```suggestion
            sourceField: "type"
            predicate: { equals: 2 }
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-equals-predicate-works-with-integer-type.yml [hunk 29]
```diff
@@ -0,0 +1,161 @@
...
+    # language=JSON
+    response: |-
+      {
+        "data": {
+          "barById": {
+            "name": "Bar A"
+          }
+        },
+        "extensions": {}
+      }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-equals-predicate-works-with-integer-type.yml:130]: This is NOT used, so remove it. The test framework should really error out if a service call isn't used.

```suggestion
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-matches-predicate-works-with-id-type.yml [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_A"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { matches: "nope" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-matches-predicate-works-with-id-type.yml:34]:
```suggestion
            sourceField: "type"
            predicate: { matches: "nope" }
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-matches-predicate-works-with-id-type.yml [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_B"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { matches: "[a-z]+_B"}
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-matches-predicate-works-with-id-type.yml:50]:
```suggestion
            sourceField: "type"
            predicate: { matches: "[a-z]+_B"}
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-matches-predicate-works-with-id-type.yml [hunk 32]
```diff
@@ -0,0 +1,161 @@
...
+    # language=JSON
+    response: |-
+      {
+        "data": {
+          "barById": {
+            "name": "Bar A"
+          }
+        },
+        "extensions": {}
+      }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-matches-predicate-works-with-id-type.yml:130]:
```suggestion
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-starts-with-predicate-works-with-id-type.yml [hunk 33]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_A"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { startsWith: "nope" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-starts-with-predicate-works-with-id-type.yml:34]:
```suggestion
            sourceField: "type"
            predicate: { startsWith: "nope" }
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-starts-with-predicate-works-with-id-type.yml [hunk 34]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_B"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { startsWith: "type" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-starts-with-predicate-works-with-id-type.yml:50]:
```suggestion
            sourceField: "type"
            predicate: { startsWith: "type" }
```

File: test/src/test/resources/fixtures/hydration/conditional-hydration-starts-with-predicate-works-with-id-type.yml [hunk 35]
```diff
@@ -0,0 +1,161 @@
...
+    # language=JSON
+    response: |-
+      {
+        "data": {
+          "barById": {
+            "name": "Bar A"
+          }
+        },
+        "extensions": {}
+      }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/conditional-hydration-starts-with-predicate-works-with-id-type.yml:130]:
```suggestion
```

File: test/src/test/resources/fixtures/hydration/hydration-with-multiple-conditions-ignores-false-conditions.yml [hunk 36]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_A"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "type_A" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/hydration-with-multiple-conditions-ignores-false-conditions.yml:34]:
```suggestion
            sourceField: "type"
            predicate: { equals: "type_A" }
```

File: test/src/test/resources/fixtures/hydration/hydration-with-multiple-conditions-ignores-false-conditions.yml [hunk 37]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_B"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "type_B" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/hydration-with-multiple-conditions-ignores-false-conditions.yml:50]:
```suggestion
            sourceField: "type"
            predicate: { equals: "type_B" }
```

File: test/src/test/resources/fixtures/hydration/hydration-with-multiple-conditions-ignores-false-conditions.yml [hunk 38]
```diff
@@ -0,0 +1,161 @@
...
+    # language=JSON
+    response: |-
+      {
+        "data": {
+          "barById": {
+            "name": "Bar A"
+          }
+        },
+        "extensions": {}
+      }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/hydration-with-multiple-conditions-ignores-false-conditions.yml:130]:
```suggestion
```

File: test/src/test/resources/fixtures/hydration/hydration-with-multiple-true-conditions-takes-first.yml [hunk 39]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_A"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { startsWith: "type" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/hydration-with-multiple-true-conditions-takes-first.yml:34]:
```suggestion
            sourceField: "type"
            predicate: { startsWith: "type" }
```

File: test/src/test/resources/fixtures/hydration/hydration-with-multiple-true-conditions-takes-first.yml [hunk 40]
```diff
@@ -0,0 +1,161 @@
...
+        arguments: [
+          {
+            name: "id" 
+            value: "BAR_B"
+          }
+        ]
+        when: {
+          result: {
+              sourceField: "type"
+              predicate: { equals: "type_B" }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/hydration-with-multiple-true-conditions-takes-first.yml:50]:
```suggestion
            sourceField: "type"
            predicate: { equals: "type_B" }
```

File: test/src/test/resources/fixtures/hydration/hydration-with-multiple-true-conditions-takes-first.yml [hunk 41]
```diff
@@ -0,0 +1,161 @@
...
+    # language=JSON
+    response: |-
+      {
+        "data": {
+          "barById": {
+            "name": "Bar B"
+          }
+        },
+        "extensions": {}
+      }
```
Comment (gnawf) [inline test/src/test/resources/fixtures/hydration/hydration-with-multiple-true-conditions-takes-first.yml:149]:
```suggestion
```

File: lib/src/main/java/graphql/nadel/engine/blueprint/NadelExecutionBlueprintFactory.kt [hunk 42]
```diff
@@ -258,26 +261,45 @@ private class Factory(
...
-            )
+            when (val expectedValue = hydration.conditionalHydration.predicate.equals) {
+                is BigInteger -> return NadelHydrationWhenCondition.LongResultEquals(
+                    fieldPath = NadelQueryPath(hydration.conditionalHydration.sourceField),
+                    value = expectedValue.longValueExact(),
+                )
+                is String -> return NadelHydrationWhenCondition.StringResultEquals(
+                    fieldPath = NadelQueryPath(hydration.conditionalHydration.sourceField),
+                    value = expectedValue
+                )
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/engine/blueprint/NadelExecutionBlueprintFactory.kt:301]: I think we should add an `else` step here and throw an error on unexpected type.

**Review by gnawf — CHANGES_REQUESTED** [review]
There's a few tests where you've left in unused service calls.

Most of the review is actionable from Github suggestions.

I think we need to add tests where the source field is null.

**Comment by sbarker2** [comment]
> There's a few tests where you've left in unused service calls.
> 

The resoning behind that was if we set up a different response for `BarA` and `BarB`, we can assert that the service call for the correct `Bar` was called. 
However I guess if we only set up a service call for the one we want then that is also fine (and less code) as it will just error/fail the test if the wrong one is called. I'll remove the unused ones.

**Review by gnawf — APPROVED** [review]
In another PR can you add a few more tests for when the field condition is null?

e.g.

```graphql
@hydrated(
  result: {
    sourceField: "type"
    predicate: { equals: "someOtherRandomTypeThatNobodyUses" }
  }
)
```

Where in the result `type` is null.
